app/shared/user-agent-http-client.js

```diff
@@ -0,0 +1,210 @@
+      return page.sessionId;
+    } else if (this.PAGE_TO_SESSION.has(page.id)) {
+      // If we have a promise to the session for this page, and the page
+      // does not have its own sessionId, we probably hit a race condition,
+      // so wait until the page has its sessionId by listening for this promise.
+      return this.PAGE_TO_SESSION.get(page.id);
+    }
+
+    // If we have neither promise nor sessionId, then something
+    // terrible has happened.
+    throw new Error(`No session mapping for page ${page.id}.`);
+  }
+
+  createSession(pageId, { scope = 0, ancestor, reason }) {
+    const requestPromise = this.uaRequest('/sessions/start', {
+      method: 'POST',
+      json: {
+        scope,
+        ancestor,
+        reason,
+      },
+    });
+
+    if (pageId) {
+      // Set the map from the page to a promise resolving to the session ID
+      // immediately for future requests.
+      this.PAGE_TO_SESSION.set(pageId, requestPromise.then(res => res.session));
+    }
+
+    return requestPromise;
+  }
+
+  async destroySession(page, { reason }) {
+    const session = await this.waitForSession(page);
+    return this.uaRequest('/sessions/end', {
+      method: 'POST',
+      json: {
+        session,
+        reason,
+      },
+    });
+  }
+
+  async createPage(page, { url, readerResult }) {
+    const session = await this.waitForSession(page);
+    return this.uaRequest('/pages/page', {
+      method: 'POST',
+      json: {
+        session,
+        url,
+        page: readerResult,
+      },
+    });
+  }
+
+  async createStar(page, { url, title }) {
+    const session = await this.waitForSession(page);
+    return this.uaRequest('/stars/star', {
+      method: 'POST',
+      json: { url, title, session },
+    });
+  }
+
+  async destroyStar(page, { url }) {
+    const session = await this.waitForSession(page);
+    return this.uaRequest('/stars/unstar', {
+      method: 'POST',
+      json: { url, session },
+    });
+  }
+
+  async createHistory(page, { url, title }) {
+    const session = await this.waitForSession(page);
+    return this.uaRequest('/visits/visit', {
+      method: 'POST',
+      json: {
+        url,
+        title,
+        session,
+      },
+    });
+  }
+
+  /**
+   * Requests that do not require a session.
+   */
+
+  async query({ text, since, limit, snippetSize }) {
+    const args = [];
+    args.push(`q=${encodeURIComponent(text)}`);
+    if (since) {
+      since = `${since}`;
+      args.push(`since=${encodeURIComponent(since)}`);
+    }
+    if (limit) {
+      limit = `${limit}`;
+      args.push(`limit=${encodeURIComponent(limit)}`);
+    }
+    if (snippetSize) {
+      snippetSize = `${snippetSize}`;
+      args.push(`snippetSize=${encodeURIComponent(snippetSize)}`);
+    }
+
+    return this.uaRequest(`/pages?${args.join('&')}`, {
+      method: 'GET',
+    });
+  }
+
+  async visited({ limit }) {
+    const args = [];
+    if (limit) {
+      limit = `${limit}`;
```

Comment:
.. and I see this was copied over from another file, feel free to leave it as-is if you want